src/python/pants/backend/python/docs/sphinx/target_types.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+from pathlib import PurePath
+from typing import Sequence
+
+from pants.core.goals.package import OutputPathField
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    InvalidFieldException,
+    MultipleSourcesField,
+    Target,
+)
+from pants.util.strutil import softwrap
+
+
+class SphinxProjectSourcesField(MultipleSourcesField):
+    # TODO: support markdown
+    default = ("conf.py", "**/*.rst")
+    expected_file_extensions = (".py", ".rst")
+    uses_source_roots = False
+
+    def validate_resolved_files(self, files: Sequence[str]) -> None:
+        super().validate_resolved_files(files)
+        py_files = [f for f in files if PurePath(f).suffix == ".py"]
+        if not py_files:
+            raise InvalidFieldException(
+                softwrap(
+                    f"""
+                    The {repr(self.alias)} field in target {self.address} must have exactly one
+                    `.py` file named `conf.py`, but no `.py` files found.
+                    """
+                )
+            )
+        if len(py_files) > 1:
```

Comment:
Is this actually an error? Why not just search for a conf.py among the sources and fail if one isn't available? Then we just have one exception to raise instead of 3. I don't know enough about sphinx to suggest that it's impossible for other .py files to be inputs, so unless you are very confident about this, it seems better to err on the side of caution. I don't see much of a downside?

Reply:
So far my understanding is yes -- and a common bug might be that people think they should include their `.py` files w/ in-line code directly in the `sphinx_project`, when really they should be using the `dependencies` field (probably, tbd)

I'm definitely willing to loosen this as I understand more.

For now, I prefer to be strict. It's easier to remove restrictions than to add them.

Reply:
It occurs to me that perhaps we should treat `config.py` like we treat `setup.py` for a `python_distribution` target. I.e., it is provided by a (transitive) dep on a python_source() target, and the target itself has no sources field. Similarly, other sources (rst files or whatever) can come from resources/files targets in the deps.

The advantage of this is regularity - there is just one target type that owns a (non-test) python source file. And config.py is a python source file. The Sphinx docs even say you "can execute arbitrarily complex code" in it.

For example, what if you want to run black on config.py? Right now that looks like it can't happen? 

I've recently been thinking that a distinction between "targets that describe sources" and "targets that describe some action you take on sources" will be important, and this would help adhere to that distinction.

Reply:
> I've recently been thinking that a distinction between "targets that describe sources" and "targets that describe some action you take on sources" will be important, and this would help adhere to that distinction.

This is a very good thing to be thinking about - it was a HUGE win when we removed `source` from `pex_binary`. It unlocked a ton of later improvements like atom targets vs generator targets.

--

Here, my instinct though is that `conf.py` really is not a normal Python file you'd use and import. For example:

* Its imports come from the Sphinx tool lockfile, not from `[python].resolves`. So, if you do the common thing of something like `from sphinx_astropy.conf import *`, then Pants will error with MyPy and Pylint, along with `[python].unowned_dependency_behavior`.
* It is nonsensical for another file to import your `conf.py`.
* We don't want to require source roots, which `python_source` does

>  I.e., it is provided by a (transitive) dep on a python_source() target, and the target itself has no sources field. Similarly, other sources (rst files or whatever) can come from resources/files targets in the deps.

Likewise, this I fear would result in much more boilerplate, especially that we could not as easily automate via `tailor`.

```python
python_sources()

resources(name="sphinx_resources", sources=["**/*.rst", "static/**"])

sphinx_project(name="sphinx", dependencies=["conf.py", ":sphinx_resources"])
```

Vs just:

```python
sphinx_project(name="sphinx")
```

--

Imagine for a second that the file was `sphinx.toml`. What would you recommend we do?

Reply:
Have another :wrench:
I would still want to run black on `conf.py`...

Reply:
I think everything you've mentioned about conf.py is true of setup.py, yet we now treat them differently, and I'm not sure that is wise. 

If it's nonsensical to import from conf.py then people won't do it and it won't be a problem. 

I feel reasonably strongly that the setup.py model has worked well, and we should follow it here.

Reply:
That boilerplate seems very easy to automate?

Reply:
> That boilerplate seems very easy to automate?

How?

Reply:
I feel I have one foot in each camp here. I'm +1 to treat the `conf.py` as a python source file for the sake of formatting and linting, but also like to keep the amount of build file targets down. I feel the presented alternative using phython_source and resources is using "low level" targets (which is doable) but that gets rather noisy and risks obscuring intent, which a more dedicated target provides more clearly.

There's also the outstanding issue of how to treat first party code, that you may want to document.

My cents.

Reply:
It just feels weird that a user can't run `./pants lint path/to/conf.py` for target modeling reasons that are obscure to them.

I mean of course really this should all work with no targets at all (in the common case), but that is a much larger discussion. That is, your statement 

> I'm thinking about ways to augment our FieldSet infrastructure so that you can say "Yeah, this is not actually a PythonSource -- but go with it still."

Is 100% right, and this should be the *norm*, not the exception. A user would see this and go, "of course its a python source, it ends with .py!" :)  In other words, we should get out of the way of what the user is obviously trying to do more than we currently are able to. But again, that is a wider discussion. 

For this specific case, doing what we already do with setup.py seems like good modeling to me, and has proven itself in that case.

Reply:
> For this specific case, doing what we already do with setup.py seems like good modeling to me, and has proven itself in that case.

With setup.py, how do you handle when it has imports? This is my biggest concern:

> Its imports come from the Sphinx tool lockfile, not from [python].resolves. So, if you do the common thing of something like from sphinx_astropy.conf import *, then Pants will error with MyPy and Pylint, along with [python].unowned_dependency_behavior.

Reply:
It does have imports! At the very least it imports from setuptools, but it can import from anywhere in principle. And that is one of the reasons why treating it as regular Python source makes sense.

Is there any rule that conf.py can't import? It sounds like it could if the user wants to do that for whatever reason?

Reply:
How about a review of your friendly neighborhood sphinx user: https://github.com/pantsbuild/pex/blob/b69c469162b6762d266e04c07643db89aa4e596a/docs/conf.py#L19-L38

Reply:
Yeah, but the fact that we have to treat BUILD files specially for, e.g., linting, is an annoyance, not a feature.  

A better example is `setup.py`, which we do treat like regular code, and that has been very helpful.

For example, what happens if `conf.py` has `import foo.util` for some utility code? Now your config consists of two files, one of which we treat specially and the other not. 

I really think python code is python code and should be treated uniformly.

Reply:
Has anyone actually looked at the Pex case? It includes a plugin as well as an import of its own source to get version. Seems way better to model off of reality rather than guesses especially when it just right over there.

Reply:
Good point. Eric, how would the current modeling handle this real-world case?